pubsub.go

```diff
@@ -240,6 +244,20 @@ func NewPubSub(ctx context.Context, h host.Host, rt PubSubRouter, opts ...Option
 	return ps, nil
 }
 
+// MsgIdFunction returns a unique ID for the passed Message, and PubSub can be customized to use any
+// implementation of this function by configuring it with the Option from WithMessageIdFn.
+type MsgIdFunction func(pmsg *pb.Message) string
+
+// WithMessageIdFn is an option to customize the way a message ID is computed for a pubsub message.
+// The default ID function is DefaultMsgIdFn (concatenate source and seq nr.),
+// but it can be customized to e.g. the hash of the message.
+func WithMessageIdFn(fn MsgIdFunction) Option {
+	return func(p *PubSub) error {
+		p.msgID = fn
+		return nil
```

Comment:
In order to resolve the tracer id function issue, let's add this code:
```
if p.tracer != nil {
  p.tracer.msgID = fn
}
```

This will allow us to pass the options in any order.

Reply:
:+1: see 6bd07a7